From 593e110f9ee5f4b28cd7b7124431714b95398b78 Mon Sep 17 00:00:00 2001
From: Yifan Zhu <fanzhuyifan@gmail.com>
Date: Sun, 23 Jun 2024 15:21:16 -0700
Subject: [PATCH] explicitly process invalid keycodes

Since dd36387c1fe704914cf41dd1e7333bac703e8ca7 in kwin, key events with
invalid key codes are also passed to kglobalacceld. The code mistakenly
triggers the corresponding modifier-only shortcut.

BUG: 489001
FIXED-IN: 6.1.2
---
 autotests/shortcutstest.cpp     | 3 +++
 src/globalshortcutsregistry.cpp | 9 +++++++++
 2 files changed, 12 insertions(+)

diff --git a/autotests/shortcutstest.cpp b/autotests/shortcutstest.cpp
index d8c0b0d..a546782 100644
--- a/autotests/shortcutstest.cpp
+++ b/autotests/shortcutstest.cpp
@@ -136,6 +136,9 @@ void ShortcutsTest::testShortcuts_data()
         << (Events() << std::make_pair(QEvent::KeyPress, Qt::Key_Control) << std::make_pair(QEvent::Wheel, 0)
                      << std::make_pair(QEvent::KeyRelease, Qt::Key_Control))
         << false;
+    QTest::newRow("mod+invalid does not trigger modifier-only single mod")
+        << QKeySequence(Qt::ControlModifier)
+        << (Events() << std::make_pair(QEvent::KeyPress, Qt::ControlModifier) << std::make_pair(QEvent::KeyRelease, Qt::ControlModifier)) << false;
 }
 
 void ShortcutsTest::testShortcuts()
diff --git a/src/globalshortcutsregistry.cpp b/src/globalshortcutsregistry.cpp
index 9b46a9d..368f412 100644
--- a/src/globalshortcutsregistry.cpp
+++ b/src/globalshortcutsregistry.cpp
@@ -419,6 +419,11 @@ bool GlobalShortcutsRegistry::keyPressed(int keyQt)
     const int key = keyQt & ~Qt::KeyboardModifierMask;
     const Qt::KeyboardModifiers modifiers = static_cast<Qt::KeyboardModifiers>(keyQt & Qt::KeyboardModifierMask);
     switch (key) {
+    case 0:
+        // Invalid key code
+        m_state = Normal;
+        processKey(0); // Reset the active sequence
+        return false;
     case Qt::Key_Shift:
     case Qt::Key_Control:
     case Qt::Key_Alt:
@@ -521,6 +526,10 @@ bool GlobalShortcutsRegistry::keyReleased(int keyQt)
     const int key = keyQt & ~Qt::KeyboardModifierMask;
     const Qt::KeyboardModifiers modifiers = static_cast<Qt::KeyboardModifiers>(keyQt & Qt::KeyboardModifierMask);
     switch (key) {
+    case 0:
+        // Invalid key code
+        m_state = Normal;
+        break;
     case Qt::Key_Super_L:
     case Qt::Key_Super_R:
     case Qt::Key_Meta:
-- 
GitLab

